640 readme cleanup (#771)
GSA/fedramp-automation: README.md

Reviewer: @kyhu65867 the README looks so much better!! In the "Build requirements are" list, can you update "Java 8+" to "Java 11+", and remove the "Python 3.9+" bullet?  I think this PR is ready once that's done.

Author: Done in commit b16d915

# Federal Risk and Authorization Management Program (FedRAMP) Automation

## Overview

> The FedRAMP Program Management Office (PMO), has drafted FedRAMP specific tooling and guidance to ensure our stakeholders can fully express and review their FedRAMP Security Authorization Packages using NIST’s [OSCAL syntax](https://pages.nist.gov/OSCAL/documentation/).

Author: I am not going to commit this suggestion but rather use it to inform my own edits.

> The FedRAMP Program Management Office (PMO) has drafted FedRAMP specific tooling and guidance to ensure our stakeholders can fully express and review their FedRAMP Security Authorization Packages using NIST’s [OSCAL SSP syntax](https://pages.nist.gov/OSCAL/documentation/).

Reviewer 2: So it has been so long I didn't realize I have not read this in years, so first off, thank you! Quick just-in-time feedback: can we say OSCAL's syntax and not _just_ SSP syntax? That is inaccurate and misleading.

```suggestion
The FedRAMP Program Management Office (PMO) has drafted FedRAMP specific tooling and guidance to ensure our stakeholders can fully express and review their FedRAMP Security Authorization Packages using NIST’s [OSCAL syntax](https://pages.nist.gov/OSCAL/documentation/).
```

> In addition to guides and examples assisting in the production of OSCAL artifacts, the FedRAMP PMO is developing software to assist in automating the review of OSCAL (Open Security Controls Assessment Language) digital authorization packages. The primary aim is to reduce manual review efforts by validating whether submissions conform to FedRAMP’s requirements. Once completed, it will help ensure that packages such as System Security Plans (SSPs) and Plans of Action and Milestones (POA&Ms) meet FedRAMP's expectations before submission, helping to streamline the review process.

Reviewer 3: I thought we weren't supposed to refer to "FedRAMP PMO" anymore. Just "FedRAMP".

> In addition to guidance and examples assisting in the production of OSCAL artifacts, the FedRAMP PMO is developing software to assist in automating the review of OSCAL (Open Security Controls Assessment Language) digital authorization packages. The primary aim is to reduce manual review efforts and timeframes by validating whether packages conform to FedRAMP’s requirements as early as possible. Once completed, FedRAMP provided tooling will help ensure that package items such as System Security Plans (SSPs) and Plans of Action and Milestones (POA&Ms) meet FedRAMP's expectations before submission, helping to streamline the review process.

Author: I like a bit of context but I shortened it down substantially

## Project Components

> | Component                                                       | Description                                                                        |

Reviewer 2: Let's be careful with the term [component](https://pages.nist.gov/OSCAL-Reference/models/v1.1.2/component-definition/), since it has a specific meaning in OSCAL.

```suggestion
| Resource                                                       | Description                                                                        |
```

Author: Question about just deleting the table entirely-- if we only have one project resource that is the validation tooling I don't think it is necessary

| ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- |

> | [FedRAMP OSCAL Validation Tooling](./tree/develop/src/validations/constraints) |  The folder that contains the validation tooling and constraints      |

Reviewer 3: Do we mean to specifically link to the constraints in the develop branch in this link?

Author: Thanks for the heads up, I removed the table and tried to make links relative. The only reason I am pointing to the develop branch is that there is no constraints folder in main, the tooling is only accessible via the develop branch

> | [FedRAMP OSCAL Validation Tooling](./tree/develop/src/validations/constraints) |  The folder that contains the validation tooling and constraints      |
> | [FedRAMP Rev 5 Releases](/tree/master/dist/content/rev5)         | The folder that contains guidance and examples related to rev5                     |

Reviewer 2: I understand the intent here, but we can use relative links without explicating the branch so, once 🤞 the latter merges we do not have to change the code before it lands.

```suggestion
| [FedRAMP OSCAL Validation Tooling](./src/validations/constraints) |  The folder that contains the validation tooling and constraints      |
| [FedRAMP Rev 5 Releases](./dist/content/rev5)         | The folder that contains guidance and examples related to rev5                     |
```

Author: Done

## FedRAMP OSCAL Validation Tooling

> Our ongoing work is focused on developing automated validation tooling using the oscal-cli to automatically check that security artifacts in a FedRAMP digital authorization package (such as an system security plan) complies with FedRAMP information and data requirements. 

Reviewer 2: Minor wording tweaks and capitalization of SSP is probably for the best.

```suggestion
Our current focus is developing validation constraints to use with the oscal-cli to automatically check that all parts of a FedRAMP digital authorization package (such as System Security Plan) meet FedRAMP's requirements before staff start a formal review. 
```

Author: done

> Our ongoing work is focused on developing automated validation tooling called oscal-cli to automatically check that FedRAMP required controls are present in security artifacts such as System Security Plans. 
> 
> As a part of this project, we are continuing to release "constraints," or automated "checks" of FedRAMP's required controls, to expand the coverage of our tooling and further automate the review of security artifacts. To learn more about installing and using our validation tooling, go [here](https://github.com/GSA/fedramp-automation/blob/develop/src/validations/constraints/README.md). 

Reviewer: Use of word "controls" might confuse users in this context as we're not talking about the security "control"  that are in the baselines. Maybe, we could say something like:

```suggestion
Our ongoing work is focused on developing automated validation tooling using the oscal-cli to automatically check that security artifacts in a FedRAMP digital authorization package (such as an system security plan) complies with FedRAMP information and data requirements. 

As a part of this project, we are continuing to release "constraints," or automated "checks" of FedRAMP's digital authorization package requirements, to expand the coverage of our tooling and further automate the review of security artifacts. To learn more about installing and using our validation tooling, go [here](https://github.com/GSA/fedramp-automation/blob/develop/src/validations/constraints/README.md). 
```

Our tooling:
- Validates OSCAL documents against FedRAMP constraints.
- Identifies compliance with FedRAMP requirements.
- Outputs a SARIF report, detailing both passed and failed validations.

This tooling is intended for use by FedRAMP OSCAL implementers and practitioners, Cloud Service Providers (CSPs), OSCAL tool developers, 3rd Party Assessment Organizations (3PAOs), and federal agencies. We welcome any and all feedback. 

## FedRAMP OSCAL Rev 5 Releases:

The FedRAMP PMO has released the following OSCAL content:

- **FedRAMP Baselines:** The FedRAMP rev 5 baselines for High, Moderate, Low, and Tailored for Low Impact-Software as a Service (LI-SaaS) in OSCAL (XML and JSON formats) are available [here](./dist/content/rev5/baselines).

- **FedRAMP OSCAL Templates:** The template files are pre-populated with FedRAMP extensions, defined-identifiers, and conformity tags where practical. They also include sample data, and are the basis for their respective guidance documents above. The FedRAMP OSCAL SSP, SAP, SAR, and POA&M template are now available [here](./documents/) in XML, JSON, and YAML formats.

> - **FedRAMP OSCAL Registry** This registry is the authoritative source for all FedRAMP extensions to the OSCAL syntax, FedRAMP-defined identifiers, and accepted values. The FedRAMP OSCAL Registry is now available [here](./dist/content/rev5/resources) in XML format.

Reviewer: These are slated for deprecation per ADR #738 .  Should we leave until then or go ahead and remove now since we're updating the page?

Author: Addressed this feedback by removing that whole section and instead pointing to the automate.fedramp.gov site

> - **FedRAMP OSCAL Registry** This registry is the authoritative source for all FedRAMP extensions to the OSCAL syntax, FedRAMP-defined identifiers, and accepted values. The FedRAMP OSCAL Registry is now available [here](./dist/content/rev5/resources) in XML format.

Reviewer 2: Do we need any of this anymore if we have the table in the earlier section? Much of this can go away or be integrated into that table, yeah?

Author: Counterpoint-- why do we need the table if none of those resources will be contained in this repository anymore?

> - **Implementation Guides:** These documents help tool developers and content authors ensure any generated OSCAL-based FedRAMP deliverabes are fully compliant with FedRAMP’s extensions, defined identifiers, conformity tags, and acceptable values. The FedRAMP OSCAL implementation guides is now available [here](./documents/) in PDF format.

Reviewer: The automate.fedramp.gov site effectively replaced the guides, so lets update this as follows:

```suggestion
- **Implementation Documentation:** The [FedRAMP Developer Hub documentation site](https://automate.fedramp.gov/documentation/) helps tool developers and content authors ensure any generated OSCAL-based FedRAMP deliverables are fully compliant with FedRAMP’s extensions, defined identifiers, and acceptable values. 
```

## Questions and Feedback

Please ask questions or provide feedback on the items above above either via email to [oscal@fedramp.gov](mailto:oscal@fedramp.gov), as a comment to an existing [issue](https://github.com/GSA/fedramp-automation/issues), or as a new [issue](https://github.com/GSA/fedramp-automation/issues).

## Dependencies and OSCAL resources

FedRAMP's work is based on NIST's [OSCAL 1.1.2](https://github.com/usnistgov/OSCAL/releases/tag/v1.1.2), and requires an understanding of the core OSCAL syntax, as well as NIST-provided resources to function correctly. As such, we have provided NIST-produced OSCAL resources below. 

**IMPORTANT**: As NIST makes minor syntax updates and releases new versions, please review [the NIST OSCAL release notes](https://pages.nist.gov/OSCAL/reference/release-notes/) in addition to guides here for more information about these changes.

> <details><summary>Resources</summary>

Reviewer 2: I appreciate we added this section to be helpful, but maybe we should just point NIST documentation and not repeat much of it. In the past, many of the things in the resource section we integrally used for our tooling. At this point, once this content merges to master, that will not be the case anymore.

Author: I removed this whole section hehe

# Federal Risk and Authorization Management Program (FedRAMP) Automation

## Overview

**If you are interested in [an overview about OSCAL](https://automate.fedramp.gov/about/) and [extensive documentation](https://automate.fedramp.gov/documentation/) on how to use it for FedRAMP's specific requirements, please visit our [Developer Hub](https://automate.fedramp.gov/).**

> FedRAMP maintains this repository with data, software, and documentation to review digital authorization packages for FedRAMP authorizations using [OSCAL](https://pages.nist.gov/OSCAL/documentation/). Our primary aim  is to reduce manual review efforts and timeframes by validating whether submissions conform to FedRAMP’s requirements. Once complete, our tooling will help ensure that packages such as System Security Plans (SSPs) and Plans of Action and Milestones (POA&Ms) meet FedRAMP's expectations before submission, streamlining the review process.

Reviewer 2: Minor spacing issue, double spacing after the phrase primary aim:

```suggestion
FedRAMP maintains this repository with data, software, and documentation to review digital authorization packages for FedRAMP authorizations using [OSCAL](https://pages.nist.gov/OSCAL/documentation/). Our primary aim is to reduce manual review efforts and timeframes by validating whether submissions conform to FedRAMP’s requirements. Once complete, our tooling will help ensure that packages such as System Security Plans (SSPs) and Plans of Action and Milestones (POA&Ms) meet FedRAMP's expectations before submission, streamlining the review process.
```

## Questions and Feedback

Please ask questions or provide feedback on the items above above either via email to [oscal@fedramp.gov](mailto:oscal@fedramp.gov), as a comment to an existing [issue](https://github.com/GSA/fedramp-automation/issues), or as a new [issue](https://github.com/GSA/fedramp-automation/issues).

## Dependencies and OSCAL resources

FedRAMP's work is based on NIST's [OSCAL 1.1.2](https://github.com/usnistgov/OSCAL/releases/tag/v1.1.2), and requires an understanding of the core OSCAL syntax, as well as NIST-provided resources to function correctly. As such, we have provided NIST-produced OSCAL resources below. 

**IMPORTANT**: As NIST makes minor syntax updates and releases new versions, please review [the NIST OSCAL release notes](https://pages.nist.gov/OSCAL/reference/release-notes/) in addition to guides here for more information about these changes.

## Developer notes

This section is for prospective contributors to our automation efforts. As an open source project, fedramp-automation welcomes contributions. To see a detailed guide for contributors, go [here](./CONTRIBUTING.md)
<details>
<summary>How to build/test our tools</summary>

### Build / test

> A top-level Makefile is provided to simplify builds.

Reviewer: @aj-stein-gsa and @wandmagic can you confirm the "build requirements"?  I think oscal-cli requires java 11+ and python 3.9 is nearing EOL...not sure we need to list python at all.

Reviewer 2: Fair enough, I would argue in hindsight that we should cut Python altogether because I forgot to recall during last review #738 that we will get rid of all relevant parts of the codebase that use Python.

- **NIST SP 800-53 & 53A Revision 5 in OSCAL:** NIST is also providing SP 800-53 and 800-53A, Revision 5 content as well as the NIST High, Moderate, and Low baselines in OSCAL (XML, JSON, and YAML formats) [here](https://github.com/usnistgov/oscal-content/tree/main/nist.gov/SP800-53/rev5).

NIST offers a complete package containing the NIST OSCAL converters, syntax validation tools, 800-53 and FedRAMP baselines content is available for download in both ZIP and BZ2 format. Visit the [NIST OSCAL Github releases page for more information](https://github.com/usnistgov/OSCAL/releases/latest).

Please ask questions or provide feedback on the above NIST dependencies either via email to [oscal@nist.gov](mailto:oscal@nist.gov), as a comment to an existing issue, or as a new issue via the [NIST OSCAL GitHub site](https://github.com/usnistgov/OSCAL/issues).

FedRAMP looks forward to receiving your comments and sharing additional progress.
</details>

## Developer notes

> This section is for prospective contributors to our automation efforts. As an open source project, fedramp-automation welcomes contributions. To see a detailed guide for contributors, go [here](https://github.com/GSA/fedramp-automation/blob/develop/CONTRIBUTING.md)

Reviewer 2: I'd also do a relative path here for similar reasons to above so the branch changes are handled for free whatever branch you are on.

```suggestion
This section is for prospective contributors to our automation efforts. As an open source project, fedramp-automation welcomes contributions. To see a detailed guide for contributors, go [here](./CONTRIBUTING.md)
```

Author: done

```

If you are developing on Windows, [msys2](https://www.msys2.org/) may be used for the required build tools (`make` and `bash`, in particular). Follow all the suggested installation steps on the msys2 home page for a complete environment. Additionally, make sure all the build requirements (above) are available on your path.
</details>

<details>
<summary>How to create a release</summary>

> ### Creating a release

Reviewer 2: I am of two minds on this, but I am not sure it matters because I doubt our own developers are reading this README. Two approaches, either:

1. Point to the checklist in [the wiki](https://github.com/GSA/fedramp-automation/wiki/Release-Prep)
2. Delete this section entirely and presume the FedRAMP Team who needs to make a release will not read this anyway.

I am all for 2, I actually set this up I believe to be transparent but to offboard to colleagues from the contractor developer team who are no longer here. Parts of this guidance was not followed, so Approach 2 will not change much.

Author: Deleted :)
